gr-blocks/lib/tcp_connection.cc

```diff
@@ -54,16 +54,25 @@ namespace gr {
     tcp_connection::send(pmt::pmt_t vector)
     {
       size_t len = pmt::blob_length(vector);
+
+      // Asio async_write() requires the buffer to remain valid until the handler is called. It will be
+      // deleted in tcp_connection::handle_write.
+      char* txbuf = new char[len];
+
+      size_t temp = 0;
+      memcpy(txbuf, pmt::uniform_vector_elements(vector, temp), len);
+
       size_t offset = 0;
-      std::vector<char> txbuf(std::min(len, d_buf.size()));
       while (offset < len) {
-        size_t send_len = std::min((len - offset), txbuf.size());
-        memcpy(&txbuf[0], pmt::uniform_vector_elements(vector, offset), send_len);
+        // Limit the size of each write() to the MTU.
+        // FIXME: Note that this has the effect of breaking a large PDU into several smaller PDUs, each
+        // containing <= MTU bytes. Is this the desired behavior?
```

Comment:
Are you talking about GNU Radio PDUs or TCP segments? TCP can always fragment and deliver a stream over multiple calls to recv(), and it is up to the receiver to reassemble any boundaries in the data.